pkg/upgrademonitor/upgrade_monitor.go

```diff
@@ -0,0 +1,266 @@
+package upgrademonitor
+
+import (
+	"context"
+	"fmt"
+
+	machineconfigurationalphav1 "github.com/openshift/client-go/machineconfiguration/applyconfigurations/machineconfiguration/v1alpha1"
+	mcfgclientset "github.com/openshift/client-go/machineconfiguration/clientset/versioned"
+	"github.com/openshift/library-go/pkg/operator/configobserver/featuregates"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	v1 "github.com/openshift/api/config/v1"
+	mcfgalphav1 "github.com/openshift/api/machineconfiguration/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+)
+
+type Condition struct {
+	State   mcfgalphav1.StateProgress
+	Reason  string
+	Message string
+}
+
+// GenerateAndApplyMachineConfigNodes takes a parent and child conditions and applies them to the given node's MachineConfigNode object
+// there are a few stipulations. 1) if the parent and child condition exactly match their currently applied statuses, no new MCN is generated
+// 2) the desiredConfig in the MCN Status will only be set once the update is proven to be compatible. Meanwhile the desired and current config in the spec react to live changes of state on the Node
+// 3) None of this will be executed unless the TechPreviewNoUpgrade featuregate is applied.
+// nolint:gocyclo
+func GenerateAndApplyMachineConfigNodes(parentCondition, childCondition *Condition, parentStatus, childStatus metav1.ConditionStatus, node *corev1.Node, mcfgClient mcfgclientset.Interface, fgAccessor featuregates.FeatureGateAccess) error {
+	if fgAccessor == nil || node == nil || parentCondition == nil || mcfgClient == nil {
+		return nil
+	}
+	fg, err := fgAccessor.CurrentFeatureGates()
+	if err != nil {
+		klog.Errorf("Could not get fg: %w", err)
+		return err
+	}
+	if fg == nil || !fg.Enabled(v1.FeatureGateMachineConfigNodes) {
+		klog.Infof("MCN Featuregate is not enabled. Please enable the TechPreviewNoUpgrade featureset to use MachineConfigNodes")
```

Comment:
The daemon seems to be spamming this: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/4012/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1729234371795226624/artifacts/e2e-gcp-op/gather-extra/artifacts/pods/openshift-machine-config-operator_machine-config-daemon-gkhst_machine-config-daemon.log

Shouldn't this have extra gating to enter this function?

Reply:
Yeah, I have this removed locally, I will push this